ext/openssl/ossl_x509cert.c

```diff
@@ -704,6 +704,94 @@ ossl_x509_eq(VALUE self, VALUE other)
     return !X509_cmp(a, b) ? Qtrue : Qfalse;
 }
 
+static VALUE
+ossl_x509_load_chained_cert_from_file(VALUE self, VALUE path)
+{
+    VALUE certs; /* Return array (Certificate object) */
+    BIO *in;
+    X509 *x509, *ca;
+    STACK_OF(X509) *certs_stack;
+    unsigned long err;
+    SSL_CTX *ctx = SSL_CTX_new(TLS_method());
+
+    in = BIO_new(BIO_s_file());
+    if (in == NULL) {
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    if (BIO_read_filename(in, StringValueCStr(path)) <= 0) {
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    x509 = PEM_read_bio_X509_AUX(in, NULL, NULL, NULL);
+    if (x509 == NULL) {
+	BIO_free(in);
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    SSL_CTX_use_certificate(ctx, x509);
+    if (ERR_peek_error() != 0) {
+        X509_free(x509);
+	BIO_free(in);
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL); /* Key/certificate mismatch doesn't imply */
+    }
+
+    /*
+     * If we could set up our certificate, now proceed to the CA
+     * certificates.
+     */
+    SSL_CTX_clear_chain_certs(ctx);
+    while ((ca = PEM_read_bio_X509(in, NULL, NULL, NULL)) != NULL) {
+        int r = SSL_CTX_add0_chain_cert(ctx, ca);
```

Comment:
Instead of adding to the context, add it to the array here.

Also need to check what is the difference between `PEM_read_bio_X509` and `PEM_read_bio_x509_AUX`.

Reply:
Thank you for your important point!

Now I can follow that[1]:
  - with `_AUX` will process a trusted X509 certificates
  - without `_AUX` will process an X509 certificates
      - will process a trusted X509 certificates but any trusted settings are discarded.

[1] https://www.openssl.org/docs/manmaster/man3/PEM_read_bio_X509.html#DESCRIPTION

As I understand, the trusted setting is only for trusted certificate [2], so using the function without `_AUX` will be fit for the purpose of this function? (Because of "discarded" meaning in [1], I may try both myself and see the differences anyway)

[2] https://www.openssl.org/docs/man1.0.2/man1/x509.html#TRUST-SETTINGS

Reply:
Hi again 😄! Could you please review my new commit d24041a? Thank you~

https://github.com/ruby/openssl/pull/361#issuecomment-619627125

Reply:
I'm continuing the implementation for also reading the DER format certificate.

For now, both PEM format and DER format could be read, but **DER format only can read the top one certificate**. I'm trying to fix this.

When you have a time, could you please review the commit https://github.com/ruby/openssl/commit/ff8572a7dd509cf345fd2654850868b08adf3279? Thank you~

https://github.com/ruby/openssl/pull/361#issuecomment-622894999